quant layer/group/instance norm: make weights and biases optional

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40055 graph mode: remove hardsigmoid op from `_single_input_general_shape_call_funcs` category
* #39413 graph mode: util for fusion of functions which require observation
* **#39203 quant layer/group/instance norm: make weights and biases optional**

Summary:

Cleans logic and test coverage for optional weights and biases for
the quantized normalization operators.  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

vkuzo committed:
quant layer/group/instance norm: make weights and biases optional

Summary:

Adds logic and test coverage for optional weights and biases for
the quantized normalization operators.  This was broken before this
PR because the `TORCH_LIBRARY` registration had these as required parameters
- removed it, and cleaned up the callsites.

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Adds logic and test coverage for optional weights and biases for
the quantized normalization operators.  This was broken before this
PR because the `TORCH_LIBRARY` registration had these as required parameters
- removed it, and cleaned up the callsites.

Note: consolidating the registrations in `native_functions.yaml` as opposed to `library.cpp`
after a discussion with @ezyang .  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -63,9 +63,6 @@ TORCH_LIBRARY(quantized, m) {
   m.def("conv3d_dilation(__torch__.torch.classes.quantized.Conv3dPackedParamsBase packed_weights) -> int[]");
   m.def("conv3d_groups(__torch__.torch.classes.quantized.Conv3dPackedParamsBase packed_weights) -> int");
   m.def("hardswish(Tensor input, float output_scale, int output_zero_point) -> Tensor");
-  m.def("group_norm(Tensor input, int num_groups, Tensor weight, Tensor bias, float eps, float output_scale, int output_zero_point) -> Tensor");
```
For understandind, why are these definitions removed here?

vkuzo replied:
yep, the main limitation is that
1. there are kernels and helper functions in c++ needed for this ops which require optional tensors
2. those kernels and helper functions are also used by floating point normalization operators (which are defined in `native_functions.yaml`)
3. the syntax for making a Tensor optional is different between this style of op registration versus `native_functions.yaml`

So, it's either choosing only `native_functions.yaml` or this style for all ops which use the kernels, or writing glue code between the different ways of handling optional tensors.  After discussing with @ezyang we decided that native_functions.yaml is the best solution for this case.

raghuramank100 commented on torch/nn/quantized/modules/normalization.py:
```diff
@@ -72,7 +72,7 @@ def __init__(self, normalized_shape, weight, bias, scale, zero_point, eps=1e-5,
 
     def forward(self, input):
-        return torch.ops.quantized.layer_norm(
+        return torch.quantized_layer_norm(
```
Wouldn't having these methods in the quantized name-space be preferrable?

vkuzo replied:
please see the comment above - the reasoning is the same here

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Adds logic and test coverage for optional weights and biases for
the quantized normalization operators.  This was broken before this
PR because the `TORCH_LIBRARY` registration had these as required parameters
- removed it, and cleaned up the callsites.

Note: consolidating the registrations in `native_functions.yaml` as opposed to `library.cpp`
after a discussion with @ezyang .  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Adds logic and test coverage for optional weights and biases for
the quantized normalization operators.  This was broken before this
PR because the `TORCH_LIBRARY` registration had these as required parameters
- removed it, and cleaned up the callsites.

Note: consolidating the registrations in `native_functions.yaml` as opposed to `library.cpp`
after a discussion with @ezyang .  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Adds logic and test coverage for optional weights and biases for
the quantized normalization operators.  This was broken before this
PR because the `TORCH_LIBRARY` registration had these as required parameters
- removed it, and cleaned up the callsites.

Note: consolidating the registrations in `native_functions.yaml` as opposed to `library.cpp`
after a discussion with @ezyang .  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Adds logic and test coverage for optional weights and biases for
the quantized normalization operators.  This was broken before this
PR because the `TORCH_LIBRARY` registration had these as required parameters
- removed it, and cleaned up the callsites.

Note: consolidating the registrations in `native_functions.yaml` as opposed to `library.cpp`
after a discussion with @ezyang .  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Adds logic and test coverage for optional weights and biases for
the quantized normalization operators.  This was broken before this
PR because the `TORCH_LIBRARY` registration had these as required parameters
- removed it, and cleaned up the callsites.

Note: consolidating the registrations in `native_functions.yaml` as opposed to `library.cpp`
after a discussion with @ezyang .  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Adds logic and test coverage for optional weights and biases for
the quantized normalization operators.  This was broken before this
PR because the `TORCH_LIBRARY` registration had these as required parameters
- removed it, and cleaned up the callsites.

Note: consolidating the registrations in `native_functions.yaml` as opposed to `library.cpp`
after a discussion with @ezyang .  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Cleans logic and test coverage for optional weights and biases for
the quantized normalization operators.  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1347,11 +1347,6 @@
     CPU: native_group_norm_backward
     CUDA: native_group_norm_backward
 
-- func: quantized_group_norm(Tensor input, int num_groups, Tensor? weight, Tensor? bias, float eps, float output_scale, int output_zero_point) -> Tensor
```
could you remove quantized_batch_norm as well?

vkuzo replied:
np.  We can do that as a separate PR, since it's not related to this diff, technically.

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Cleans logic and test coverage for optional weights and biases for
the quantized normalization operators.  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

jerryzh168 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -108,8 +108,10 @@
-    # Quantize functions should not record gradients
+    # Quantize tensor functions should not record gradients
     'quantize_per_tensor', 'quantize_per_channel',
+    # Operations on quantized tensors do not have derivatives, for now
+    'quantized_layer_norm', 'quantized_group_norm', 'quantized_instance_norm',
```
wait, aren't these removed?

vkuzo replied:
this was a holdover from before ezyang's fix, fixed it, thanks for the callout

jerryzh168 commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -83,6 +83,9 @@
     ('quantized::linear_dynamic', datetime.date(2020, 6, 1)),
     ('quantized::linear_relu', datetime.date(2020, 6, 1)),
     ('quantized::linear', datetime.date(2020, 6, 1)),
+    ('quantized::layer_norm', datetime.date(2020, 6, 30)),
+    ('quantized::group_norm', datetime.date(2020, 6, 30)),
+    ('quantized::instance_norm', datetime.date(2020, 6, 30)),
```
I didn't see changes to these functions in the PR, why are they here?

vkuzo replied:
fixed, thanks!

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Cleans logic and test coverage for optional weights and biases for
the quantized normalization operators.  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Cleans logic and test coverage for optional weights and biases for
the quantized normalization operators.  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks good

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Cleans logic and test coverage for optional weights and biases for
the quantized normalization operators.  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]

vkuzo committed:
Update on "quant layer/group/instance norm: make weights and biases optional"


Summary:

Cleans logic and test coverage for optional weights and biases for
the quantized normalization operators.  

Test Plan:

```
python test/test_quantization.py TestQuantizedOps.test_qlayer_norm
python test/test_quantization.py TestQuantizedOps.test_group_norm
python test/test_quantization.py TestQuantizedOps.test_instance_norm
python test/test_quantization.py TestStaticQuantizedModule.test_layer_norm
python test/test_quantization.py TestStaticQuantizedModule.test_group_norm
python test/test_quantization.py TestStaticQuantizedModule.test_instance_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_layer_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_group_norm
python test/test_quantization.py TestQuantizeScriptPTSQOps.test_instance_norm
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21885259](https://our.internmc.facebook.com/intern/diff/D21885259)

[ghstack-poisoned]
